pkg/intermediate/aggregate.go

```diff
@@ -524,59 +548,101 @@ func (a *AggregationProcess) aggregateRecords(incomingRecord, existingRecord ent
+				if dstExistingIeWithValue, _, exist := existingRecord.GetInfoElementWithValue(antreaDestinationStatsElements[i]); exist {
+					existingVal := dstExistingIeWithValue.GetUnsigned64Value()
 					if !isDelta {
-						existingIeWithValue.SetUnsigned64Value(incomingVal)
+						dstExistingIeWithValue.SetUnsigned64Value(incomingVal)
+						switch antreaDestinationStatsElements[i] {
+						case "octetTotalCountFromDestinationNode":
+							totalCountDiff = incomingVal - existingVal
+						case "reverseOctetTotalCountFromDestinationNode":
+							reverseTotalCountDiff = incomingVal - existingVal
+						}
 					} else {
-						existingIeWithValue.SetUnsigned64Value(incomingVal + existingIeWithValue.GetUnsigned64Value())
+						dstExistingIeWithValue.SetUnsigned64Value(incomingVal + existingVal)
 					}
 				} else {
 					return fmt.Errorf("element does not exist in the record: %v", antreaDestinationStatsElements[i])
 				}
 			}
+			// Update the corresponding common element in statsElement list.
+			commonExistingIeWithValue, _, _ := existingRecord.GetInfoElementWithValue(element)
+			if isLatest {
+				if !isDelta {
+					if commonExistingIeWithValue.GetUnsigned64Value() < incomingVal {
+						commonExistingIeWithValue.SetUnsigned64Value(incomingVal)
+					}
+				} else {
+					if fillSrcStats {
+						srcIe, _, _ := existingRecord.GetInfoElementWithValue(antreaSourceStatsElements[i])
+						commonExistingIeWithValue.SetUnsigned64Value(srcIe.GetUnsigned64Value())
+					}
+					if fillDstStats {
+						dstIe, _, _ := existingRecord.GetInfoElementWithValue(antreaDestinationStatsElements[i])
+						commonExistingIeWithValue.SetUnsigned64Value(dstIe.GetUnsigned64Value())
+					}
+				}
+			}
 		} else {
 			return fmt.Errorf("element with name %v in statsElements not present in the incoming record", element)
 		}
 	}
+
+	if flowEndSecondsDiff == 0 {
+		return nil
+	}
+	// Update the throughput fields: throughput = (octetTotalCount - prev_octetTotalCount) / (flowEndSeconds - prev_flowEndSeconds)
+	throughput := totalCountDiff / uint64(flowEndSecondsDiff)
+	if fillSrcStats {
+		ie, _, _ := existingRecord.GetInfoElementWithValue("throughputFromSourceNode")
+		ie.SetUnsigned64Value(throughput)
+	}
+	if fillDstStats {
+		ie, _, _ := existingRecord.GetInfoElementWithValue("throughputFromDestinationNode")
+		ie.SetUnsigned64Value(throughput)
+	}
+	ie, _, _ := existingRecord.GetInfoElementWithValue("throughput")
+	ie.SetUnsigned64Value(throughput)
+
+	// Update the reverseThroughput fields
+	reverseThroughput := reverseTotalCountDiff / uint64(flowEndSecondsDiff)
+	if fillSrcStats {
+		ie, _, _ := existingRecord.GetInfoElementWithValue("reverseThroughputFromSourceNode")
+		ie.SetUnsigned64Value(reverseThroughput)
+	}
+	if fillDstStats {
+		ie, _, _ := existingRecord.GetInfoElementWithValue("reverseThroughputFromDestinationNode")
+		ie.SetUnsigned64Value(reverseThroughput)
+	}
```

Comment:
I don't think we should hardcode creation of fields here. Instead, we can take these fields as input like how we handle stats. We can add input like `timestampFields` to take all flowStartSeconds, flowEndSeconds, flowEndSecondsFromSourceNode etc. related fields.

Reply:
things are a little bit different for throughput elements.. As for the stats, we have all the elements in statsElementList in the flow record, len(statsElementList) = len(antreaSourceStatsElements) = len(antreaDestinationStatsElements), we can take advantage of the index, and expand to source and destination fields. So this is the challenge.

I also agree this section can be optimized. I'll think about how to re-structure it.